src/command.rs

```diff
@@ -77,255 +77,268 @@ pub async fn init(kind: PackageType, name: Option<PackageName>) -> miette::Resul
     Ok(())
 }
 
-/// Adds a dependency to this project
-pub async fn add(registry: RegistryUri, dependency: &str) -> miette::Result<()> {
-    let lower_kebab = |c: char| (c.is_lowercase() && c.is_ascii_alphabetic()) || c == '-';
-
-    let (repository, dependency) = dependency
-        .trim()
-        .split_once('/')
-        .ok_or_else(|| miette!("locator {dependency} is missing a repository delimiter"))?;
-
-    ensure!(
-        repository.chars().all(lower_kebab),
-        "repository {repository} is not in kebab case"
-    );
-
-    let repository = repository.into();
-
-    let (package, version) = dependency
-        .split_once('@')
-        .ok_or_else(|| miette!("dependency specification is missing version part: {dependency}"))?;
-
-    let package = package
-        .parse::<PackageName>()
-        .wrap_err(miette!("invalid package name: {package}"))?;
-
-    let version = version
-        .parse::<VersionReq>()
-        .into_diagnostic()
-        .wrap_err(miette!("not a valid version requirement: {version}"))?;
-
-    let mut manifest = Manifest::read().await?;
-
-    manifest
-        .dependencies
-        .push(Dependency::new(registry, repository, package, version));
-
-    manifest
-        .write()
-        .await
-        .wrap_err(miette!("failed to write `{MANIFEST_FILE}`"))
-}
+impl Context {
+    /// Adds a dependency to this project
+    pub async fn add(
```

Comment:
Can we inverse this? I think it should take the context, rather than being implemented _on_ the context.

Reply:
The usual way in Rust to structure code like this is by creating contexts and using methods on them. See `reqwest::Client` for example. This allows for dot-chaining (`Context::current().some_method()`), generates useful documentation (*what methods can I call on this context*) and allows for good IDE developer experience (like auto-complete).

Is there a reason we want to do it differently in this project? I can put it into a separate PR if there is some discussion/strategizing that is needed here.

Reply:
I think we loose the notion of having one function per buffrs command, we are very likely to take the context in different places (like eg. - the resolver) aswell and (this is a personal opinion, but maybe relevant:) 
I think the "context" is really just grouped data in memory which you typically want to pass around together, rather than a proper concept which can do things (in your code a context _can_ (semantically) _add_ a dependency) 

Rather than the function that adds a dependency, needing a context to operate in. 

To me these do the same but have two very different semantic properties attached, that's why I marked this. Also it feels odd to have certain commands that don't need a context detached from the rest of the commands. The init command is now in a weird place - and there could be a couple more commands like the init one

Reply:
I am happy to invert this if you would like — it is your project after all. 

However, I fundamentally disagree with some of your takes. What I will do is summarize my position and thought process behind this change, that way you have all of the information. If after reading this, you still disagree, then we will do it the way you have proposed.

## Notion of one function per command

> I think we loose the notion of having one function per buffrs command, we are very likely to take the context in different places

I would say, having one function per buffrs command is explicitly an anti-goal. Functions as the primary interface for running `buffrs` are (in my opinion/experience) not the "Rustacean" way of doing things. I would not allow such code to be merged into a project of mine.

This is best explained if you think about this in terms of a programmer who might want to consume `buffrs` as a library, perhaps by building something on top of it. What does this programmer need to do? 

- Open a new `Context`, this can fail if you are not inside a valid `git` repository.
- Find out what you can do with this context (actions as methods)

I will give you a real-life example of this pattern. When you interact with `git` from within Rust:

- You create a new [`Repository`](https://docs.rs/gix/latest/gix/struct.Repository.html), this can fail
- You can find out what actions you can take on this repository because they are exposed as methods

This example in [`gix`](https://docs.rs/gix/latest/gix/struct.Repository.html) is a good example because it resembles closely what we are building here. A low-level abstraction that can be used to interact with a `buffrs` project.

In some languages, [such as Zig (iirc)](https://github.com/ziglang/zig/issues/1170), there is no difference at all between a *method* and a *function*, it is just syntactic sugar. In Rust, there is. It is more of a semantic one, as any method you can also call like `Repository::do_something(&value)` rather than `instance.value()`.

But the difference is: when we define something as a method, semantically it means that this is an action that can be taken on that value. For example, when you *install* a buffrs project, the syntax should be `Context::current().install()`. 

When you use functions, it becomes very difficult to figure out what actions can be taken. This is because in Rustdoc, you cannot filter functions for *all functions that take a value X as parameter*. And the reason you cannot do this is because you do not need to because you are supposed to use `impl` blocks to *associate* functions to values.

So, in my opinion: we specifically don't want to have one *function* per buffrs command, but rather one *method*. It is fine if there are some places where we pass a `Context` by reference. But for the main entrypoint (public API), it should be methods. Functions should not be publicly exposed.

## Grouped data

> I think the "context" is really just grouped data in memory which you typically want to pass around together, rather than a proper concept which can do things. To me these do the same but have two very different semantic properties attached, that's why I marked this.

Everything is fundamentally just *grouped data*, right. But this type is not just a convenience type that groups some arguments so you don't have to repeat them. It carries a significant semantic meaning with it.

If you have an value of type `Context`, it means that you have made sure that at the paths specified within it, there really is a `buffrs` project. You need to do that step to be able to call any action on the `buffrs` project. So, semantically the `Context` states: *at some point in the flow of this program have we made sure that this is in fact a valid `buffrs` project*.

Of course, we are interacting with the filesystem, which is globally shared and mutable state. So we can never be fully certain that you do not delete the folder during the invocation of the command. But our responsibility is to, at least once in the execution, ascertain ourselves that we are in a valid buffrs project so that we can emit a useful error message if this is not the case.

So: in my opinion, this `Context` *is* in fact a proper context with properties that must be upheld. It is an abstraction.

## Init command

> Also it feels odd to have certain commands that don't need a context detached from the rest of the commands. The init command is now in a weird place - and there could be a couple more commands like the init one.

This is not by accident. The code is merely a reflection of the reality.

Using buffrs looks like this:

```mermaid
graph LR
    empty
    empty-->buffrs_init
    buffrs_init-->buffrs_add
    buffrs_init-->buffrs_install
    buffrs_init-->buffrs_remove
    buffrs_init-->buffrs_package
    buffrs_init-->buffrs_publish
    buffrs_init-->buffrs_generate
```

To summarize: there are two distrinct sets of commands here. There are the commands which expect a properly initialized `buffrs` project to be in place (add, install, remove, package, publish, generate, ...) and there is the set of commands that do not (init).

I would perhaps clean this up and make the `init` be an associated function of the `Context` struct and have it return a context. 

## Summary

In my opinion, the changes I made here make sense and overall improve the code base. I also feel that this is not a personal opinion but a style that is widely mirrored in the Rust community. 

If, after this explanation, you agree with these choices we can leave these in. If you disagree, that is okay too. In that case, I will take those changes out of this PR.